File: python_solutions/DynamicProgramming/Best_Time_to_Buy_and_Sell_Stock_IV.py
class Solution:
    def maxProfit(self, k, prices):
        """
        :type k: int
        :type prices: List[int]
        :rtype: int
        """
        if not prices:
            return 0
        if k > len(prices)//2:
            return self.quickSolve(prices)
        
        buy = [prices[0]]*(k+1)
        sell = [0]*(k+1)
        
        for p in prices:
            for i in range(1, k+1):
                buy[i] = min(buy[i], p-sell[i-1])
                sell[i] = max(sell[i], p-buy[i])
        
        return sell[k]
    
    def quickSolve(self, prices):
        profit = 0
        for i in range(1, len(prices)):
            if prices[i] > prices[i-1]:
                profit += prices[i] - prices[i-1]
                
        return profit
            

File: cpp_solutions/DynamicProgramming/Best_Time_to_Buy_and_Sell_Stock_IV.cpp
/*
* Say you have an array for which the ith element is the price of a given stock on day i.
*
* Design an algorithm to find the maximum profit. You may complete at most k transactions.
*
*/
// Notes: hard to think, 
// local vector, and global vector to keep i-th transaction result
// keep updating local and global.

class Solution {
public:
    int maxProfit(int k, vector<int> &prices) {
        if (prices.empty())
            return 0;

        int ans = 0;
        if (k >= prices.size() / 2)
        {
            for (int i = 1; i < prices.size(); ++i) {
                if (prices[i] - prices[i - 1] > 0) {
                    ans += prices[i] - prices[i - 1];
                }
            }
        } else {
            vector<int> local(k+1);
            vector<int> global(k+1);
            
            for (int i = 0; i < prices.size() - 1; ++i) {
                int increase = prices[i + 1] - prices[i];
    
                for (int j = k; j >= 1; --j) {
                    local[j] = max(global[j - 1] + max(increase, 0), local[j] + increase);
                    global[j] = max(global[j], local[j]);
                }
            }
            ans = global[k];
        }
        return ans;
    }
   
};